src/data/nat/basic.lean

```diff
@@ -1062,8 +1069,34 @@ lemma with_bot.add_eq_one_iff : ∀ {n m : with_bot ℕ}, n + m = 1 ↔ (n = 0 
 by apply nat.less_than_or_equal.rec h0; exact h1
 
 @[elab_as_eliminator]
-lemma decreasing_induction {P : ℕ → Prop} (h : ∀n, P (n+1) → P n) {m n : ℕ} (nm : m ≤ n) (hP : P n) :
-  P m :=
-by { induction nm with n nm ih, exact hP, exact ih (h _ hP) }
+def decreasing_induction {P : ℕ → Sort*} (h : ∀n, P (n+1) → P n) {m n : ℕ} (mn : m ≤ n)
+  (hP : P n) : P m :=
+le_rec_on mn (λ k ih hsk, ih $ h k hsk) (λ h, h) hP
+
+lemma decreasing_induction_self {P : ℕ → Sort*} (h : ∀n, P (n+1) → P n) {n : ℕ} (nn : n ≤ n)
```

Comment:
`simp`

Reply:
Also `n le n` is an argument here, instead of just being proved in the statement. What is the reason for this?

Reply:
I mostly copied the statement from `le_rec_on_self`. I guess it doesn't really matter because we have definitional proof irrelevance. Do `rw` and `simp` still work if they have to match `le_refl n =?= h`?

Reply:
I tried it, and `rw` matches more easily when the proof on `n \le n` is a variable. In fact, changing this breaks the (current) proof of `decreasing_induction_trans`.